docs/RFCS/20230605_workload_level_index_recommendation.md

```diff
@@ -0,0 +1,348 @@
+`(k)`. In total, we will merge all the storing parts in one node and find the shallowest leaf node to put them if there 
+is no covering path is found. Next I will explain why we use this simple idea. First of all, the space cost of putting 
+them together storing in one leaf one and splitting them into multiple groups and storing them in multiple leaf nodes is
+almost the same. Next, if we accidentally split the storing part of one original index into two, then the corresponding
+query cannot be optimized. In addition, we do not want to make the index too large (the size of key and value). Thus, we
+decide to merge all the storing parts in one node together and put them in one leaf node, the shallowest one.
+
+To sum up, we build a trie tree on the indexed columns from previous index recommendations amd attach the storing parts
+to each tree node. Then, select all the leaf nodes to build representative indexes. For each node with attached storing
+parts, they will be merged. Next, if we do not find a path from the one node to the leaf node in its subtree, meaning
+that there does not exist an index storing all the columns we need, then we will choose the shallowest leaf nodes to 
+mount all the storing columns. 
+
+## Drop index
+Let us consider all the `DROP INDEX` for the next step. It only exists in the index recommendation for replace where the 
+index to drop shares the same index columns as the index to create, but the storing columns is a subset compared to the 
+index to create. If we think about it from the perspective of trie tree, all the dropped indexes are covered by all the 
+indexes represented by the leaf nodes. Therefore, we can drop all the `Drop`-based indexes.
+
+## User Interface
+
+### How to use
+Initially, several built-in function will be provided command line interface. `workload_index_recs`, 
+`workload_index_recs(timestamp)`, `workload_index_recs(budget)`, and 
+`workload_index_recs(timestamp, budget)`. User can enter `SELECT workload_index_recs();`, then they will see
+the workload level index recommendations. For the parameter, the type of the `timestamp` can be TIMESTAMP/TIMESTAMPZ. 
+We will use it as a filter and only consider those indexes after the timestamp from the table 
+`system.statement_statistics`. The `budget` is a string like "15GB", "500MB", "1TB" meaning the space the user can 
+afford so that we will apply some algorithm to reduce the space cost, you can find details in the cost optimization 
+section later.
+
+### Insight delivered to users
+Since we have the statistics for the frequency about the queries, so that we can deliver the "usefulness" for each 
+index, lik, the number of query types can be optimized, the number of sql statements can be optimized (some type of 
+queries may be executed many times). In addition, we can also deliver the space cost for all the index recommendations.
+
+## Future work
```

Comment:
Another future work item you should consider mentioning is that eventually we'd like the algorithm to consider the overhead of index maintenance (not just storage cost). If there are many mutation statements in the workload (INSERT, UPDATE, DELETE, UPSERT), then indexes that contain columns that are updated will increase the cost of the mutation statements.